openmc/deplete/chain.py

```diff
@@ -0,0 +1,786 @@
+loaded from an .xml file and all the nuclides are linked together.
+"""
+
+from io import StringIO
+from itertools import chain
+import math
+import re
+from collections import OrderedDict, defaultdict
+from collections.abc import Mapping, Iterable
+from numbers import Real
+from warnings import warn
+
+from openmc.checkvalue import check_type, check_greater_than
+from openmc.data import gnd_name, zam
+from .nuclide import FissionYieldDistribution
+
+# Try to use lxml if it is available. It preserves the order of attributes and
+# provides a pretty-printer by default. If not available,
+# use OpenMC function to pretty print.
+try:
+    import lxml.etree as ET
+    _have_lxml = True
+except ImportError:
+    import xml.etree.ElementTree as ET
+    _have_lxml = False
+import scipy.sparse as sp
+
+import openmc.data
+from openmc._xml import clean_indentation
+from .nuclide import Nuclide, DecayTuple, ReactionTuple
+
+
+# tuple of (reaction name, possible MT values, (dA, dZ)) where dA is the change
+# in the mass number and dZ is the change in the atomic number
+_REACTIONS = [
+    ('(n,2n)', set(chain([16], range(875, 892))), (-1, 0)),
+    ('(n,3n)', {17}, (-2, 0)),
+    ('(n,4n)', {37}, (-3, 0)),
+    ('(n,gamma)', {102}, (1, 0)),
+    ('(n,p)', set(chain([103], range(600, 650))), (0, -1)),
+    ('(n,a)', set(chain([107], range(800, 850))), (-3, -2))
+]
+
+
+__all__ = ["Chain"]
+
+def replace_missing(product, decay_data):
+    """Replace missing product with suitable decay daughter.
+
+    Parameters
+    ----------
+    product : str
+        Name of product in GND format, e.g. 'Y86_m1'.
+    decay_data : dict
+        Dictionary of decay data
+
+    Returns
+    -------
+    product : str
+        Replacement for missing product in GND format.
+
+    """
+    # Determine atomic number, mass number, and metastable state
+    Z, A, state = openmc.data.zam(product)
+    symbol = openmc.data.ATOMIC_SYMBOL[Z]
+
+    # Replace neutron with proton
+    if Z == 0 and A == 1:
+        return 'H1'
+
+    # First check if ground state is available
+    if state:
+        product = '{}{}'.format(symbol, A)
+
+    # Find isotope with longest half-life
+    half_life = 0.0
+    for nuclide, data in decay_data.items():
+        m = re.match(r'{}(\d+)(?:_m\d+)?'.format(symbol), nuclide)
+        if m:
+            # If we find a stable nuclide, stop search
+            if data.nuclide['stable']:
+                mass_longest_lived = int(m.group(1))
+                break
+            if data.half_life.nominal_value > half_life:
+                mass_longest_lived = int(m.group(1))
+                half_life = data.half_life.nominal_value
+
+    # If mass number of longest-lived isotope is less than that of missing
+    # product, assume it undergoes beta-. Otherwise assume beta+.
+    beta_minus = (mass_longest_lived < A)
+
+    # Iterate until we find an existing nuclide
+    while product not in decay_data:
+        if Z > 98:
```

Comment:
Optional: I would define `98` here as a class / file level constant, or leave a comment here explaining it's significance.